infrastructure/tari_util/src/chacha20.rs

```diff
@@ -0,0 +1,269 @@
+fn rotl(a: u32, n: u32) -> u32 {
+    (a << n) | (a >> (32 - n))
+}
+
+/// Perform a single chacha quarter round at the provided indices in the state
+fn quarter_round(state: &mut [u32; 16], a_index: usize, b_index: usize, c_index: usize, d_index: usize) {
+    let mut a = Wrapping(state[a_index]);
+    let mut b = Wrapping(state[b_index]);
+    let mut c = Wrapping(state[c_index]);
+    let mut d = Wrapping(state[d_index]);
+    a += b;
+    d = Wrapping(rotl((d ^ a).0, 16));
+    c += d;
+    b = Wrapping(rotl((b ^ c).0, 12));
+    a += b;
+    d = Wrapping(rotl((d ^ a).0, 8));
+    c += d;
+    b = Wrapping(rotl((b ^ c).0, 7));
+    state[a_index] = a.0;
+    state[b_index] = b.0;
+    state[c_index] = c.0;
+    state[d_index] = d.0;
+}
+
+/// Construct a chacha block by performing a number of column and diagonal quarter round operations
+fn chacha20_block(state: &[u32; 16]) -> [u32; 16] {
+    let mut working_state = state.clone();
+    for _iter in 0..10 {
+        // 20 total => odd and even round performed for every iteration
+        // Odd round
+        quarter_round(&mut working_state, 0, 4, 8, 12);
+        quarter_round(&mut working_state, 1, 5, 9, 13);
+        quarter_round(&mut working_state, 2, 6, 10, 14);
+        quarter_round(&mut working_state, 3, 7, 11, 15);
+        // Even round
+        quarter_round(&mut working_state, 0, 5, 10, 15);
+        quarter_round(&mut working_state, 1, 6, 11, 12);
+        quarter_round(&mut working_state, 2, 7, 8, 13);
+        quarter_round(&mut working_state, 3, 4, 9, 14);
+    }
+    let mut output: [u32; 16] = [0; 16];
+    for i in 0..output.len() {
+        output[i] = (Wrapping(working_state[i]) + Wrapping(state[i])).0;
+    }
+    (output)
+}
+
+/// Construct an initial state from a 128-bit constant, 256-bit key, 96-bit nonce and a 32-bit block counter
+fn construct_state(key: &[u8; 32], nonce: &[u32; 3], counter: u32) -> [u32; 16] {
+    let constant: [u8; 16] = [101, 120, 112, 97, 110, 100, 32, 51, 50, 45, 98, 121, 116, 101, 32, 107]; // 0x61707865, 0x3320646e, 0x79622d32, 0x6b206574
+    let mut state_bytes = constant.to_vec(); // 128 bit
+    state_bytes.extend_from_slice(key); // 256-bit
+    state_bytes.extend_from_slice(&u32_to_u8_le(counter)); // 32-bit
+    state_bytes.extend_from_slice(&u32_to_u8_le(nonce[0])); // 96-bit
+    state_bytes.extend_from_slice(&u32_to_u8_le(nonce[1]));
+    state_bytes.extend_from_slice(&u32_to_u8_le(nonce[2]));
+    // Convert [u8;64] to [u32;16]
+    let mut curr_bytes: [u8; 4] = [0; 4];
+    let mut state: [u32; 16] = [0; 16];
+    for i in 0..state.len() {
+        let byte_index = i * 4;
+        curr_bytes.copy_from_slice(&state_bytes[byte_index..(byte_index + 4)]);
+        state[i] = u8_to_u32_le(&curr_bytes);
+    }
+    (state)
+}
+
+/// Generate a keystream consisting of a number of chacha blocks
+fn chacha20_cipher_keystream(key: &[u8; 32], nonce: &[u32; 3], block_count: usize) -> Vec<u8> {
+    let bytes_per_block: usize = 64;
```

Comment:
nit
```suggestion
    const BYTES_PER_BLOCK: usize = 64;
```
compiler might not inline this with `let`, and/or needlessly put this on the stack

Reply:
Change applied